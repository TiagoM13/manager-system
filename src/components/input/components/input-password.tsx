import React from 'react';
import { Controller, FieldValues } from 'react-hook-form';

import { Eye, EyeSlash } from '@phosphor-icons/react';

import { ErrorMessage } from '@/components/error-message';

import { InputPasswordProps } from './interfaces';

export const InputPassword = <T extends FieldValues>({
  label,
  required = false,
  className = '',
  control,
  name,
  defaultValue,
  error,
  loading = false,
  stylesRightButton = '',
  ...props
}: InputPasswordProps<T>) => {
  const [showPassword, setShowPassword] = React.useState(false);

  const togglePasswordVisibility = () => {
    setShowPassword((prevState) => !prevState);
  };

  const icon = (
    <>
      {showPassword ? (
        <EyeSlash
          data-testid="icon-eye-slash"
          weight="bold"
          color="white"
          className="size-5"
        />
      ) : (
        <Eye
          data-testid="icon-eye"
          weight="bold"
          color="white"
          className="size-5"
        />
      )}
    </>
  );

  return (
    <div className="w-full">
      <label htmlFor={name} className="block text-sm text-slate-600 mb-2">
        {label} {required && <span className="text-red-500">*</span>}
      </label>

      <div className="relative flex items-center">
        {control ? (
          <Controller
            name={name}
            control={control}
            defaultValue={defaultValue || ''}
            render={({ field }) => (
              <input
                {...field}
                {...props}
                type={showPassword ? 'text' : 'password'}
                className={`${className} w-full rounded-s-md border-y border-l border-slate-400 py-2 px-4 outline-offset-0 outline-sky-500 text-sm text-slate-600 disabled:opacity-60`}
              />
            )}
          />
        ) : (
          <input
            {...props}
            type={showPassword ? 'password' : 'text'}
            className={`${className} w-full rounded-s-md border-y border-l border-slate-400 py-2 px-4 outline-offset-0 outline-sky-500 text-sm text-slate-600 disabled:opacity-60`}
          />
        )}

        <button
          type="button"
          data-testid="icon-button"
          onClick={togglePasswordVisibility}
          className={`${stylesRightButton} flex flex-wrap items-center rounded-r-lg py-2.5 px-2.5 h-full bg-sky-600 hover:bg-sky-500 disabled:bg-sky-700 outline-sky-500`}
          disabled={loading}
        >
          {icon}
        </button>
      </div>

      <ErrorMessage error={error?.message} />
    </div>
  );
};
