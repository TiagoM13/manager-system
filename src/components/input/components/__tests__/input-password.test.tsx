import React from 'react';

import { fireEvent, render } from '@testing-library/react';

import { InputPassword } from '../input-password';

describe('<InputPassword/>', () => {
  const label = 'Sua senha';
  const placeholder = 'Escreva sua senha';

  test('should render correctly component', () => {
    const {
      getByText,
      getByTestId,
      getByPlaceholderText,
      container,
      asFragment,
    } = render(
      <InputPassword
        name="password"
        label={label}
        placeholder={placeholder}
        required
      />,
    );

    const iconEye = getByTestId('icon-eye');
    const element = container.querySelector('.text-red-500');

    expect(getByText(label)).toBeInTheDocument();
    expect(getByPlaceholderText(placeholder)).toBeInTheDocument();
    expect(element).toBeInTheDocument();
    expect(iconEye).toBeTruthy();

    expect(asFragment()).toMatchSnapshot();
  });

  test('should change password input icon', () => {
    const { getByTestId, getByPlaceholderText } = render(
      <InputPassword name="password" placeholder={placeholder} />,
    );

    const input = getByPlaceholderText(placeholder) as HTMLInputElement;
    const iconEye = getByTestId('icon-eye');

    expect(iconEye).toBeTruthy();
    expect(input.type).toBe('text');

    const button = getByTestId('icon-button');

    fireEvent.click(button);

    const iconEyeSlash = getByTestId('icon-eye-slash');

    expect(iconEyeSlash).toBeTruthy();
    expect(iconEye).not.toBeInTheDocument();
    expect(input.type).toBe('password');
  });
});
