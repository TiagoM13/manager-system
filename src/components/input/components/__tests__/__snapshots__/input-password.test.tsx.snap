// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<InputPassword/> should render correctly component 1`] = `
<DocumentFragment>
  <div
    class="w-full"
  >
    <label
      class="block text-sm text-slate-600 mb-2"
      for="password"
    >
      Sua senha 
      <span
        class="text-red-500"
      >
        *
      </span>
    </label>
    <div
      class="relative flex items-center"
    >
      <input
        class=" w-full rounded-s-md border-y border-l border-slate-400 py-2 px-4 outline-sky-500 text-sm text-slate-600 disabled:opacity-60"
        placeholder="Escreva sua senha"
        type="text"
      />
      <button
        class=" flex flex-wrap items-center rounded-r-lg py-2.5 px-2.5 h-full bg-sky-600 hover:bg-sky-500 disabled:bg-sky-700 outline-sky-500"
        data-testid="icon-button"
        type="button"
      >
        <svg
          class="size-5"
          data-testid="icon-eye"
          fill="white"
          height="1em"
          viewBox="0 0 256 256"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M251,123.13c-.37-.81-9.13-20.26-28.48-39.61C196.63,57.67,164,44,128,44S59.37,57.67,33.51,83.52C14.16,102.87,5.4,122.32,5,123.13a12.08,12.08,0,0,0,0,9.75c.37.82,9.13,20.26,28.49,39.61C59.37,198.34,92,212,128,212s68.63-13.66,94.48-39.51c19.36-19.35,28.12-38.79,28.49-39.61A12.08,12.08,0,0,0,251,123.13Zm-46.06,33C183.47,177.27,157.59,188,128,188s-55.47-10.73-76.91-31.88A130.36,130.36,0,0,1,29.52,128,130.45,130.45,0,0,1,51.09,99.89C72.54,78.73,98.41,68,128,68s55.46,10.73,76.91,31.89A130.36,130.36,0,0,1,226.48,128,130.45,130.45,0,0,1,204.91,156.12ZM128,84a44,44,0,1,0,44,44A44.05,44.05,0,0,0,128,84Zm0,64a20,20,0,1,1,20-20A20,20,0,0,1,128,148Z"
          />
        </svg>
      </button>
    </div>
  </div>
</DocumentFragment>
`;
